### DISCUSSION TITLE
Cannot change data directory location

### DISCUSSION THREAD

**Discussion by dravadhis** [TriliumNext/Trilium#6158] [Q&A] [answered]
### Trilium Version

0.50.2

### What operating system are you using?

Ubuntu

### What is your setup?

Local (no sync)

### Operating System Version

Ubunutu 20.04 LTS

### Description

So I want to change the data directory location for my Trilium desktop setup. I followed the [instructions](https://github.com/zadam/trilium/wiki/Data-directory#changing-the-location-of-data-directory) as given in the documentation. However the data directory location is still not changing.

The exact steps I perform are:
1. Make sure trilium is not running. (If it is running, I close it completely)
2. In a terminal, I run the command: `export TRILIUM_DATA_DIR=home/<usr_name>/<some_path>`.
3. I then quit the terminal. Now I start Trilium.

However the data directory value doesn't change. 

I have performed a similar change successfully on a Windows setup (Windows 10 Home, Trilium Desktop Local with No sync) by following the last comment in this [discussion](https://github.com/zadam/trilium/issues/250). 

My final aim is that I should be able to use the same `trilium-data` folder for my Windows and Ubunutu dual-boot setup.

**Comment by sigaloid** [comment]
After step 2 and before step 3, can you try running `echo $TRILIUM_DATA_DIR`?

**Comment by dravadhis** [comment]
I did try `echo TRILIUM_DATA_DIR` and 'env | grep TRILIUM' after step 2 and before 3. The value was set correctly. I have tried different location values, updating, reinstalling but nothing seemed to work.

**Comment by sigaloid** [comment]
If you edit the `trilium.desktop` file, you can add the environment variable before the `trilium` invocation.

1. `sudo nano /usr/share/applications/trilium.desktop`
2. Edit the line that says `Exec=trilium %u`
3. Add `env TRILIUM_DATA_DIR=*dir*` before `trilium` like so: 
4. `Exec=env TRILIUM_DATA_DIR=/tmp/asdf trilium %U`
5. Ctrl + X, Y, Enter to save
That environment variable will then be loaded when launching Trilium.

**Comment by zadam** [comment]
You write:

> In a terminal, I run the command: export TRILIUM_DATA_DIR=home/<usr_name>/<some_path>.
> I then quit the terminal. Now I start Trilium.

But that's not how env. variables work in Linux - if you close the terminal, the env. variable you've set and exported disappears. To be active, you need to start trilium from terminal where you exported the variable.

Alternatively you can set and export the variable in e.g. `.bashrc` file.

**Comment by dravadhis** [comment]
Ok so I first followed what @zadam mentioned. (It completely slipped my mind that env. variables set using `export` disappear on closing the terminal). So I started a new terminal, set the variable as required, inspected its value using `env | grep TRILIUM` and then started trilium from the same terminal. It was still not changed.

Next I set the variable permanently in the `.bashrc` file. This also didn't work.

Finally I went ahead with what @sigaloid said. Still got no success.

**Comment by sigaloid** [comment]
What happens if you launch it directly from the terminal (env var before `trilium`)? Try that, see if it changes it, if not there might be an error message of help.

**Comment by dravadhis** [comment]
> What happens if you launch it directly from the terminal (env var before `trilium`)? Try that, see if it changes it, if not there might be an error message of help.

This worked just now.

**Comment by dravadhis** [comment]
How should I make this permanent? Starting with the env variable everytime is a bit annoying (though I'll still do it if there's no solution. The app is just that good ;)).

**Comment by dravadhis** [comment]
Ok so I started Trilium today morning and realised that it's now working with `TRILIUM_DATA_DIR` value set to what I want.
So I performed a test.
1. I first backed up my `trilium-data` folder. Then I deleted the `TRILIUM_DATA_DIR` environment variable from the `.bashrc` file and uninstalled trilium.
2. Then I installed Trilium (0.50.3) and set the `TRILIUM_DATA_DIR` in `.bashrc` file to point at my backed up copy of `trilium-data` dir. However the data directory value didn't change.
3. So I started a new terminal and launched Trilium directly from the terminal with env var before Trilium as mentioned by @sigaloid previously.

> What happens if you launch it directly from the terminal (env var before `trilium`)? Try that, see if it changes it, if not there might be an error message of help.

4. The directory value was changed. Then I quit Trilium and started it again in a normal way (directly from the Application Menu). Surprisingly the data directory value was updated. I was now able to start Trilium normally with the updated value.

I don't know why it worked this way, but I think listing the steps might help someone facing a similar problem.

Thanks @sigaloid @zadam for your help. Although the above is a roundabout method, it's worth it for an application as good as Trilium.

**Comment by Nriver** [comment, marked as answer]
`env` is used for single command, it's a one-off, only affect one line.
`export` will change the environment variable for you whole terminal session. If you close the terminal, it goes away.

`.bashrc` is the file that will be loaded automatically when you open a new terminal. So, if you put `export TRILIUM_DATA_DIR=home/<usr_name>/<some_path>` to the end of your `.bashrc` file. The setting should be available for every new terminal session now.
